nalgebra-macros/src/lib.rs

```diff
@@ -321,3 +322,398 @@ pub fn point(stream: TokenStream) -> TokenStream {
+    }
+}
+
+/// Constructs a new matrix by concatenating the given matrices.
+///
+/// **Note: Requires the `macros` feature to be enabled (enabled by default)**.
+///
+/// The syntax is similar to the [`matrix!`] and [`dmatrix!`]) macros. However the elements should
+/// be of type `&Matrix` or be one of the litterals `0` or `1`. The elements of type `&Matrix` are
+/// concatenated as expected. The litteral `0` is expanded to the zero. The litteral `1` is
+/// expanded to the identity matrix. Note that at least one element in each row and column must be
+/// an expression of type `&Matrix`.
+///
+/// All elements in the same row need to have the same number of rows and simillary for the
+/// elements in the same column. This is checked at compile time as long as all elements have
+/// dimensions of type `Const<N>`. If one or more elements have dimensions of type `Dynamic` then
+/// it is checked at runtime that all dimensions match.
+///
+/// If at least one element of a row has `Const<N>` number of rows then that row has a whole will
+/// have `Const<N>` number of rows. However if at least one row has has `Dynamic` number of rows
+/// then the entire matrix will have `Dynamic` number of rows. Same for columns.
+///
+/// # Examples
+///
+/// ```
+/// use nalgebra::{cat, matrix};
+///
+/// let a = matrix![1,2;3,4;];
+///
+/// let m1 = cat![
+///     &a, 0;
+///     0, &matrix![5,6;7,8;];
+/// ];
+///
+/// let m2 = matrix![
+///     1,2,0,0;
+///     3,4,0,0;
+///     0,0,5,6;
+///     0,0,7,8;
+/// ];
+///
+/// assert_eq!(m1, m2);
+/// ```
+///
+/// ```
+/// use nalgebra::{cat, matrix, Matrix5x6};
+///
+/// let a: Matrix5x6<_> = cat![
+///     1, &matrix![1;2], 0;
+///     0, 1, &matrix![3,4;5,6;];
+///     &matrix![7,8,9;], 0, 0;
+/// ];
+///
+/// let b = matrix![
+///     1, 0, 0, 1, 0, 0;
+///     0, 1, 0, 2, 0, 0;
+///     0, 0, 0, 1, 3, 4;
+///     0, 0, 0, 0, 5, 6;
+///     7, 8, 9, 0, 0, 0;
+/// ];
+///
+/// assert_eq!(a, b);
+/// ```
+///
+/// ```
+/// use nalgebra::{cat, matrix, dmatrix, DMatrix};
+///
+/// let a: DMatrix<_> = cat![
+///     &dmatrix![1,2;3,4;], 0;
+///     0, &matrix![5,6;7,8;];
+/// ];
+///
+/// let b = dmatrix![
+///     1,2,0,0;
+///     3,4,0,0;
+///     0,0,5,6;
+///     0,0,7,8;
+/// ];
+///
+/// assert_eq!(a, b);
+/// ```
+///
+/// ```
+/// use nalgebra::{cat, matrix, dmatrix, Matrix4};
+///
+/// let a: Matrix4<_> = cat![
+///     &dmatrix![1,2;3,4;], &matrix![5,6;7,8;];
+///     &matrix![9,10;11,12;], 0;
+/// ];
+///
+/// let b = matrix![
+///     1,2,5,6;
+///     3,4,7,8;
+///     9,10,0,0;
+///     11,12,0,0;
+/// ];
+///
+/// assert_eq!(a, b);
+/// ```
+#[proc_macro]
+pub fn cat(stream: TokenStream) -> TokenStream {
+    let matrix = parse_macro_input!(stream as Matrix);
+    proc_macro::TokenStream::from(cat_impl("__11f075cdd4a86538", matrix))
+}
+
+#[allow(clippy::too_many_lines)]
+fn cat_impl(prefix: &str, matrix: Matrix) -> TokenStream2 {
+    let n_macro_rows = matrix.nrows();
```

Comment:
`prefix` is so that we can have long complicated names for the internal variables that won't shadow external variables. This is really only relevant for the `expr` and `expr_shape` variables as they are introduced in an order where this could potentially happen.